watertap/core/zero_order_diso.py

```diff
@@ -0,0 +1,244 @@
+# "https://github.com/watertap-org/watertap/"
+#
+###############################################################################
+"""
+This module contains the methods for constructing the material balances for
+zero-order double-input/single-output (DISO) unit models (i.e. units with two inlets and single
+outlet where composition changes, such as a generic bioreactor).
+"""
+
+import idaes.logger as idaeslog
+from idaes.core.util import get_solver
+import idaes.core.util.scaling as iscale
+
+from pyomo.environ import NonNegativeReals, Var, units as pyunits
+
+# Some more information about this module
+__author__ = "Adam Atia"
+
+# Set up logger
+_log = idaeslog.getLogger(__name__)
+
+
+def build_diso(self):
+    """
+    Helper method for constructing material balances for zero-order type models
+    with DISO behavior.
+
+    Three StateBlocks are added with two corresponding Ports:
+        * properties_in1 --> inlet1
+        * properties_in2 --> inlet2
+        * properties_treated ---> treated
+
+
+    Two additional variables are added:
+        * recovery_frac_mass_H2O (indexed by time)
+        * removal_frac_mass_solute (indexed by time and solute)
+
+    Two additional constraints are added to represent the material balances
+        * water_recovery_equation (indexed by time)
+        * solute_treated_equation (indexed by time and solute)
+
+    This method also sets private attributes on the unit model with references
+    to the appropriate initialization and scaling methods to use and to return
+    the inlet volumetric flow rate.
+    """
+    self._has_recovery_removal = True
+    self._initialize = initialize_diso
+    self._scaling = calculate_scaling_factors_diso
+
+    # Create state blocks for inlet and outlets
+    tmp_dict = dict(**self.config.property_package_args)
+    tmp_dict["has_phase_equilibrium"] = False
+    tmp_dict["defined_state"] = True
+
+    self.properties_in1 = self.config.property_package.build_state_block(
+        self.flowsheet().time,
+        doc="Material properties at inlet 1",
+        default=tmp_dict)
+
+    self.properties_in2 = self.config.property_package.build_state_block(
+        self.flowsheet().time,
+        doc="Material properties at inlet 2",
+        default=tmp_dict)
+
+    tmp_dict_2 = dict(**tmp_dict)
+    tmp_dict_2["defined_state"] = False
+
+    self.properties_treated = \
+        self.config.property_package.build_state_block(
+            self.flowsheet().time,
+            doc="Material properties of treated water",
+            default=tmp_dict_2)
+
+    # Create Ports
+    self.add_port("inlet1", self.properties_in1, doc="Inlet port 1")
+    self.add_port("inlet2", self.properties_in2, doc="Inlet port 2")
+    self.add_port("treated",
+                  self.properties_treated,
+                  doc="Treated water outlet port")
+
+    # Add performance variables
+    self.recovery_frac_mass_H2O = Var(
+        self.flowsheet().time,
+        initialize=0.8,
```

Comment:
Should we be assigning defaults here? Seeing as we expect the user to load data later (or provide values), it might make sense to leave the initial value as `None` as that might make it easier for us to debug later.